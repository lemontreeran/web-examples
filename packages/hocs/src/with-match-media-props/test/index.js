import React from 'react'
import { mount } from 'enzyme'

const Target = () => null

describe('withMatchMediaProps', () => {
  let mockJson2mq = null
  let withMatchMediaProps = null

  beforeEach(() => {
    mockJson2mq = jest.fn()

    jest.mock('json2mq', () => mockJson2mq)
    jest.resetModules()

    withMatchMediaProps = require('../src/').default
  })

  afterAll(() => {
    jest.unmock('json2mq')
  })

  describe('`window.matchMedia` is supported', () => {
    let originMatchMedia = null

    beforeAll(() => {
      originMatchMedia = global.matchMedia
    })

    beforeEach(() => {
      global.matchMedia = () => {}
    })

    afterAll(() => {
      global.matchMedia = originMatchMedia
    })

    it('should just pass props through', () => {
      const EnhancedTarget = withMatchMediaProps()(Target)
      const wrapper = mount(
        <EnhancedTarget a={1} b={2} c={3} />
      )

      expect(wrapper.find(Target)).toMatchSnapshot()
    })

    it('should pass query object to json2mq', () => {
      global.matchMedia = jest.fn(() => ({
        addListener: () => {},
        removeListener: () => {},
        matches: true
      }))

      const EnhancedTarget = withMatchMediaProps({
        test: {
          maxWidth: 300
        }
      })(Target)

      mount(
        <EnhancedTarget />
      )

      expect(mockJson2mq.mock.calls).toMatchSnapshot()
    })

    it('should set initial state and provide props with matched queries', () => {
      global.matchMedia = jest.fn(() => ({
        addListener: () => {},
        removeListener: () => {},
        matches: true
      }))

      const EnhancedTarget = withMatchMediaProps({
        test: {
          maxWidth: 300
        }
      })(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper.find(Target)).toMatchSnapshot()
    })

    it('should subscribe on mount and unsubscribe on unmount', () => {
      const mockAddListener = jest.fn()
      const mockRemoveListener = jest.fn()

      global.matchMedia = jest.fn(() => ({
        addListener: mockAddListener,
        removeListener: mockRemoveListener
      }))

      const EnhancedTarget = withMatchMediaProps({
        test: {
          maxWidth: 300
        }
      })(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(mockAddListener.mock.calls).toMatchSnapshot()
      wrapper.unmount()
      expect(mockRemoveListener.mock.calls).toMatchSnapshot()
      expect(mockRemoveListener).toHaveBeenCalledWith(mockAddListener.mock.calls[0][0])
    })

    it('should change state and provide props when query has been matched', () => {
      const mockAddListener = jest.fn()

      global.matchMedia = jest.fn(() => ({
        addListener: mockAddListener,
        removeListener: () => {},
        matches: false
      }))

      const EnhancedTarget = withMatchMediaProps({
        test: {
          maxWidth: 300
        }
      })(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      mockAddListener.mock.calls[0][0]({ matches: true })
      wrapper.update()

      expect(wrapper.find(Target)).toMatchSnapshot()
    })

    describe('display name', () => {
      const origNodeEnv = process.env.NODE_ENV

      afterAll(() => {
        process.env.NODE_ENV = origNodeEnv
      })

      it('should wrap display name in non-production env', () => {
        process.env.NODE_ENV = 'test'

        const EnhancedTarget = withMatchMediaProps()(Target)
        const wrapper = mount(
          <EnhancedTarget />
        )

        expect(wrapper).toMatchSnapshot()
      })

      it('should not wrap display name in production env', () => {
        process.env.NODE_ENV = 'production'

        const EnhancedTarget = withMatchMediaProps()(Target)
        const wrapper = mount(
          <EnhancedTarget />
        )

        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('`window.matchMedia` is not supported', () => {
    it('should just pass Target component through', () => {
      const EnhancedTarget = withMatchMediaProps()(Target)
      const wrapper = mount(
        <EnhancedTarget a={1} b={2} c={3} />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
