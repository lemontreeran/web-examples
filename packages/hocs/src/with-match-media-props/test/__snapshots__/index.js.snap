// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMatchMediaProps \`window.matchMedia\` is not supported should just pass Target component through 1`] = `
<Target
  a={1}
  b={2}
  c={3}
/>
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported display name should not wrap display name in production env 1`] = `
<WithMatchMediaProps>
  <Target />
</WithMatchMediaProps>
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported display name should wrap display name in non-production env 1`] = `
<withMatchMediaProps(Target)>
  <Target />
</withMatchMediaProps(Target)>
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported should change state and provide props when query has been matched 1`] = `
<Target
  test={true}
/>
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported should just pass props through 1`] = `
<Target
  a={1}
  b={2}
  c={3}
/>
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported should pass query object to json2mq 1`] = `
Array [
  Array [
    Object {
      "maxWidth": 300,
    },
  ],
]
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported should set initial state and provide props with matched queries 1`] = `
<Target
  test={true}
/>
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported should subscribe on mount and unsubscribe on unmount 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`withMatchMediaProps \`window.matchMedia\` is supported should subscribe on mount and unsubscribe on unmount 2`] = `
Array [
  Array [
    [Function],
  ],
]
`;
